crate/utils/src/crypto/hybrid_encryption/ecies.rs

```diff
@@ -0,0 +1,204 @@
+use num_bigint_dig::algorithms::idiv_ceil;
+#[cfg(feature = "fips")]
+use openssl::nid::Nid;
+use openssl::{
+    bn::BigNumContext,
+    ec::{EcGroupRef, EcKey, EcPoint, EcPointRef, PointConversionForm},
+    hash::{Hasher, MessageDigest},
+    pkey::{PKey, Private, Public},
+    symm::{decrypt_aead, encrypt_aead, Cipher},
+};
+
+#[cfg(feature = "fips")]
+use crate::kmip_utils_bail;
+use crate::{
+    crypto::symmetric::{AES_256_GCM_IV_LENGTH, AES_256_GCM_KEY_LENGTH, AES_256_GCM_MAC_LENGTH},
+    error::KmipUtilsError,
+};
+
+/// Derive initialization vector from recipient public key `Q` and ephemeral
+/// public key `R` using SHAKE128-128.
+#[allow(non_snake_case)]
+fn ecies_get_iv(
+    Q: &EcPointRef,
+    R: &EcPointRef,
+    curve: &EcGroupRef,
+) -> Result<Vec<u8>, KmipUtilsError> {
+    let mut ctx = BigNumContext::new_secure()?;
+    let Q_vec = Q.to_bytes(curve, PointConversionForm::COMPRESSED, &mut ctx)?;
```

Comment:
Which call is made twice ?

Reply:
cf the binding crate:
```rust
    pub fn to_bytes(
        &self,
        group: &EcGroupRef,
        form: PointConversionForm,
        ctx: &mut BigNumContextRef,
    ) -> Result<Vec<u8>, ErrorStack> {
        unsafe {
            let len = ffi::EC_POINT_point2oct(
                group.as_ptr(),
                self.as_ptr(),
                form.0,
                ptr::null_mut(),
                0,
                ctx.as_ptr(),
            );
            if len == 0 {
                return Err(ErrorStack::get());
            }
            let mut buf = vec![0; len];
            let len = ffi::EC_POINT_point2oct(
                group.as_ptr(),
                self.as_ptr(),
                form.0,
                buf.as_mut_ptr(),
                len,
                ctx.as_ptr(),
            );
            if len == 0 {
                Err(ErrorStack::get())
            } else {
                Ok(buf)
            }
        }
    }
```

Reply:
I'm really not sure what's the issue here ? to_bytes() call is made only once per point ?

Reply:
There is no issue with your code.

I just remarked that there was two call to the OpenSSL conversion in the binding you are correctly using. This means conversions into bytes may be twice as slow as the C conversion. Maybe a quick check of the OpenSSL binding crate could be done on a later time to ensure there is no way to avoid this double call.